File: packages/testing/src/mcp/hints.ts
/**
 * @fileoverview Agent hints system for @lushly-dev/afd-testing
 *
 * Provides structured hints to help AI agents interpret command results
 * and determine next actions.
 */

import type { CommandResult } from '@lushly-dev/afd-core';
import type { TestReport } from '../types/report.js';

// ============================================================================
// Types
// ============================================================================

/**
 * Hints to help AI agents interpret results and decide next actions.
 */
export interface AgentHints {
	/** Should the agent retry this operation? */
	shouldRetry: boolean;

	/** Related commands to consider */
	relatedCommands: string[];

	/** Suggested next actions */
	nextSteps: string[];

	/** Confidence in result interpretation (0-1) */
	interpretationConfidence: number;

	/** Steps with low confidence that may need human review */
	lowConfidenceSteps?: number[];

	/** Commands that have no test coverage */
	untestedCommands?: string[];

	/** Error codes encountered */
	errorCodes?: string[];
}

/**
 * Command result with agent hints added.
 */
export interface AgentEnhancedResult<T> extends CommandResult<T> {
	_agentHints: AgentHints;
}

// ============================================================================
// Hint Generation
// ============================================================================

/**
 * Generate agent hints for a command result.
 */
export function generateAgentHints<T>(commandName: string, result: CommandResult<T>): AgentHints {
	const hints: AgentHints = {
		shouldRetry: shouldRetryCommand(result),
		relatedCommands: getRelatedCommands(commandName, result),
		nextSteps: suggestNextSteps(commandName, result),
		interpretationConfidence: calculateInterpretationConfidence(result),
	};

	// Add error codes if present
	if (!result.success && result.error) {
		hints.errorCodes = [result.error.code];
	}

	return hints;
}

/**
 * Generate agent hints specific to test reports.
 */
export function generateTestReportHints(report: TestReport): AgentHints {
	const { summary, scenarios } = report;

	const hints: AgentHints = {
		shouldRetry: false,
		relatedCommands: [],
		nextSteps: [],
		interpretationConfidence: summary.passRate,
	};

	// Check for failures
	const failedScenarios = scenarios.filter((s) => s.outcome === 'fail' || s.outcome === 'error');

	if (failedScenarios.length > 0) {
		hints.nextSteps.push(
			`Review ${failedScenarios.length} failed scenario(s)`,
			'Run with --verbose for detailed step output'
		);
		hints.relatedCommands.push('scenario-suggest --context failed');

		// Extract error types from failed steps
		const errorTypes = new Set<string>();
		for (const scenario of failedScenarios) {
			for (const step of scenario.stepResults) {
				if (step.error?.type) {
					errorTypes.add(step.error.type);
				}
			}
		}
		if (errorTypes.size > 0) {
			hints.errorCodes = Array.from(errorTypes);
		}
	}

	// Check for steps with expectation mismatches (lower confidence results)
	const mismatchSteps: number[] = [];
	let stepIndex = 0;
	for (const scenario of scenarios) {
		for (const step of scenario.stepResults) {
			if (step.error?.type === 'expectation_mismatch') {
				mismatchSteps.push(stepIndex);
			}
			stepIndex++;
		}
	}
	if (mismatchSteps.length > 0) {
		hints.lowConfidenceSteps = mismatchSteps;
		hints.nextSteps.push(
			`${mismatchSteps.length} step(s) have expectation mismatches - consider reviewing assertions`
		);
	}

	// Suggest based on pass rate
	if (summary.passRate >= 0.95) {
		hints.nextSteps.push('All tests passing - safe to proceed');
	} else if (summary.passRate >= 0.8) {
		hints.nextSteps.push('Most tests passing - review failures before proceeding');
	} else {
		hints.shouldRetry = true;
		hints.nextSteps.push('Significant failures detected - fix issues before continuing');
	}

	return hints;
}

/**
 * Generate agent hints for coverage results.
 */
export function generateCoverageHints(
	_testedCommands: string[],
	untestedCommands: string[],
	coveragePercent: number
): AgentHints {
	const hints: AgentHints = {
		shouldRetry: false,
		relatedCommands: [],
		nextSteps: [],
		interpretationConfidence: 0.95,
	};

	if (untestedCommands.length > 0) {
		hints.untestedCommands = untestedCommands;
		hints.relatedCommands.push('scenario-create --template crud');
		hints.nextSteps.push(
			`${untestedCommands.length} command(s) have no test coverage`,
			'Consider using scenario-create to generate test templates'
		);

		// Prioritize high-value commands
		const priorityCommands = untestedCommands.filter(
			(cmd) => cmd.includes('.create') || cmd.includes('.delete')
		);
		if (priorityCommands.length > 0) {
			hints.nextSteps.push(`Priority: Test ${priorityCommands.join(', ')} (mutation commands)`);
		}
	}

	if (coveragePercent >= 90) {
		hints.nextSteps.push('Excellent coverage - focus on edge cases');
	} else if (coveragePercent >= 70) {
		hints.nextSteps.push('Good coverage - add tests for remaining commands');
	} else {
		hints.nextSteps.push('Low coverage - prioritize adding test scenarios');
	}

	return hints;
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Determine if a command should be retried based on result.
 */
function shouldRetryCommand<T>(result: CommandResult<T>): boolean {
	if (result.success) return false;
	if (!result.error) return false;

	// Check explicit retryable flag
	if (result.error.retryable === true) return true;
	if (result.error.retryable === false) return false;

	// Check suggestion text for retry hints
	const suggestion = result.error.suggestion?.toLowerCase() ?? '';
	if (
		suggestion.includes('try again') ||
		suggestion.includes('retry') ||
		suggestion.includes('temporary')
	) {
		return true;
	}

	// Check error codes that are typically transient
	const transientCodes = ['TIMEOUT', 'CONNECTION_ERROR', 'RATE_LIMITED', 'SERVICE_UNAVAILABLE'];
	if (transientCodes.includes(result.error.code)) {
		return true;
	}

	return false;
}

/**
 * Get related commands based on the executed command.
 */
function getRelatedCommands<T>(commandName: string, result: CommandResult<T>): string[] {
	const related: string[] = [];

	// Parse command category (e.g., "scenario" from "scenario-list")
	const [category] = commandName.split('-');

	// Suggest related scenario commands
	if (category === 'scenario') {
		if (commandName === 'scenario-list') {
			related.push('scenario-evaluate', 'scenario-coverage');
		} else if (commandName === 'scenario-evaluate') {
			related.push('scenario-coverage', 'scenario-suggest');
		} else if (commandName === 'scenario-coverage') {
			related.push('scenario-suggest', 'scenario-create');
		} else if (commandName === 'scenario-create') {
			related.push('scenario-evaluate');
		}
	}

	// On failure, suggest diagnostic commands
	if (!result.success) {
		if (!related.includes('scenario-suggest')) {
			related.push('scenario-suggest --context failed');
		}
	}

	return related;
}

/**
 * Suggest next steps based on command result.
 */
function suggestNextSteps<T>(commandName: string, result: CommandResult<T>): string[] {
	const steps: string[] = [];

	if (result.success) {
		// Success suggestions
		if (commandName === 'scenario-list') {
			steps.push('Run scenario-evaluate to execute listed scenarios');
		} else if (commandName === 'scenario-create') {
			steps.push('Edit the generated scenario to add specific test cases');
			steps.push('Run scenario-evaluate to test the new scenario');
		} else if (commandName === 'scenario-evaluate') {
			steps.push('Run scenario-coverage to check test coverage');
		} else if (commandName === 'scenario-coverage') {
			steps.push('Use scenario-suggest to find gaps');
			steps.push('Create scenarios for untested commands');
		}
	} else {
		// Failure suggestions
		const error = result.error;
		if (error?.suggestion) {
			steps.push(error.suggestion);
		}

		if (error?.code === 'PARSE_ERROR') {
			steps.push('Check scenario YAML syntax');
		} else if (error?.code === 'FILE_NOT_FOUND') {
			steps.push('Verify the scenario file path exists');
		} else if (error?.code === 'VALIDATION_ERROR') {
			steps.push('Review the scenario schema requirements');
		}
	}

	return steps;
}

/**
 * Calculate confidence in how accurately we can interpret the result.
 */
function calculateInterpretationConfidence<T>(result: CommandResult<T>): number {
	let confidence = 0.9; // Base confidence

	// Lower confidence for errors without clear codes
	if (!result.success && !result.error?.code) {
		confidence -= 0.2;
	}

	// Higher confidence when reasoning is provided
	if (result.reasoning) {
		confidence += 0.05;
	}

	// Higher confidence when sources are provided
	if (result.sources && result.sources.length > 0) {
		confidence += 0.05;
	}

	// Clamp to 0-1 range
	return Math.max(0, Math.min(1, confidence));
}

// ============================================================================
// Result Enhancement
// ============================================================================

/**
 * Enhance a command result with agent hints.
 */
export function enhanceWithAgentHints<T>(
	commandName: string,
	result: CommandResult<T>
): AgentEnhancedResult<T> {
	return {
		...result,
		_agentHints: generateAgentHints(commandName, result),
	};
}

File: python/src/afd/testing/hints.py
"""
Agent hints system for AFD testing.

Provides structured hints to help AI agents interpret command results
and determine next actions.

Port of packages/testing/src/mcp/hints.ts
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Generic, TypeVar

T = TypeVar("T")


@dataclass
class AgentHints:
	"""Hints to help AI agents interpret results and decide next actions."""

	should_retry: bool
	related_commands: list[str]
	next_steps: list[str]
	interpretation_confidence: float
	low_confidence_steps: list[int] | None = None
	untested_commands: list[str] | None = None
	error_codes: list[str] | None = None


@dataclass
class AgentEnhancedResult:
	"""Command result with agent hints added."""

	success: bool
	data: Any | None = None
	error: dict[str, Any] | None = None
	_agent_hints: AgentHints | None = None


def generate_agent_hints(command_name: str, result: dict[str, Any]) -> AgentHints:
	"""Generate agent hints for a command result."""
	hints = AgentHints(
		should_retry=_should_retry_command(result),
		related_commands=_get_related_commands(command_name, result),
		next_steps=_suggest_next_steps(command_name, result),
		interpretation_confidence=_calculate_interpretation_confidence(result),
	)

	if not result.get("success") and result.get("error"):
		error = result["error"]
		code = error.get("code") if isinstance(error, dict) else getattr(error, "code", None)
		if code:
			hints.error_codes = [code]

	return hints


def generate_test_report_hints(report: dict[str, Any]) -> AgentHints:
	"""Generate agent hints specific to test reports."""
	summary = report.get("summary", {})
	scenarios = report.get("scenarios", [])

	pass_rate = summary.get("pass_rate", 0.0)

	hints = AgentHints(
		should_retry=False,
		related_commands=[],
		next_steps=[],
		interpretation_confidence=pass_rate,
	)

	failed_scenarios = [
		s for s in scenarios if s.get("outcome") in ("fail", "error")
	]

	if failed_scenarios:
		hints.next_steps.append(f"Review {len(failed_scenarios)} failed scenario(s)")
		hints.next_steps.append("Run with --verbose for detailed step output")
		hints.related_commands.append("scenario-suggest --context failed")

		error_types: set[str] = set()
		for scenario in failed_scenarios:
			for step in scenario.get("step_results", []):
				step_error = step.get("error")
				if step_error and step_error.get("type"):
					error_types.add(step_error["type"])
		if error_types:
			hints.error_codes = list(error_types)

	mismatch_steps: list[int] = []
	step_index = 0
	for scenario in scenarios:
		for step in scenario.get("step_results", []):
			if step.get("error", {}).get("type") == "expectation_mismatch":
				mismatch_steps.append(step_index)
			step_index += 1
	if mismatch_steps:
		hints.low_confidence_steps = mismatch_steps
		hints.next_steps.append(
			f"{len(mismatch_steps)} step(s) have expectation mismatches - consider reviewing assertions"
		)

	if pass_rate >= 0.95:
		hints.next_steps.append("All tests passing - safe to proceed")
	elif pass_rate >= 0.8:
		hints.next_steps.append("Most tests passing - review failures before proceeding")
	else:
		hints.should_retry = True
		hints.next_steps.append("Significant failures detected - fix issues before continuing")

	return hints


def generate_coverage_hints(
	tested: list[str],
	untested: list[str],
	coverage_percent: float,
) -> AgentHints:
	"""Generate agent hints for coverage results."""
	hints = AgentHints(
		should_retry=False,
		related_commands=[],
		next_steps=[],
		interpretation_confidence=0.95,
	)

	if untested:
		hints.untested_commands = untested
		hints.related_commands.append("scenario-create --template crud")
		hints.next_steps.append(f"{len(untested)} command(s) have no test coverage")
		hints.next_steps.append("Consider using scenario-create to generate test templates")

		priority_commands = [
			cmd for cmd in untested if ".create" in cmd or ".delete" in cmd
		]
		if priority_commands:
			hints.next_steps.append(
				f"Priority: Test {', '.join(priority_commands)} (mutation commands)"
			)

	if coverage_percent >= 90:
		hints.next_steps.append("Excellent coverage - focus on edge cases")
	elif coverage_percent >= 70:
		hints.next_steps.append("Good coverage - add tests for remaining commands")
	else:
		hints.next_steps.append("Low coverage - prioritize adding test scenarios")

	return hints


def enhance_with_agent_hints(command_name: str, result: dict[str, Any]) -> dict[str, Any]:
	"""Enhance a command result with agent hints."""
	hints = generate_agent_hints(command_name, result)
	return {**result, "_agent_hints": hints}


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _should_retry_command(result: dict[str, Any]) -> bool:
	"""Determine if a command should be retried based on result."""
	if result.get("success"):
		return False

	error = result.get("error")
	if not error:
		return False

	if isinstance(error, dict):
		retryable = error.get("retryable")
		suggestion = (error.get("suggestion") or "").lower()
		code = error.get("code", "")
	else:
		retryable = getattr(error, "retryable", None)
		suggestion = (getattr(error, "suggestion", "") or "").lower()
		code = getattr(error, "code", "")

	if retryable is True:
		return True
	if retryable is False:
		return False

	if any(kw in suggestion for kw in ("try again", "retry", "temporary")):
		return True

	transient_codes = {"TIMEOUT", "CONNECTION_ERROR", "RATE_LIMITED", "SERVICE_UNAVAILABLE"}
	if code in transient_codes:
		return True

	return False


def _get_related_commands(command_name: str, result: dict[str, Any]) -> list[str]:
	"""Get related commands based on the executed command."""
	related: list[str] = []
	category = command_name.split("-")[0] if "-" in command_name else command_name

	if category == "scenario":
		if command_name == "scenario-list":
			related.extend(["scenario-evaluate", "scenario-coverage"])
		elif command_name == "scenario-evaluate":
			related.extend(["scenario-coverage", "scenario-suggest"])
		elif command_name == "scenario-coverage":
			related.extend(["scenario-suggest", "scenario-create"])
		elif command_name == "scenario-create":
			related.append("scenario-evaluate")

	if not result.get("success"):
		if "scenario-suggest --context failed" not in related:
			related.append("scenario-suggest --context failed")

	return related


def _suggest_next_steps(command_name: str, result: dict[str, Any]) -> list[str]:
	"""Suggest next steps based on command result."""
	steps: list[str] = []

	if result.get("success"):
		if command_name == "scenario-list":
			steps.append("Run scenario-evaluate to execute listed scenarios")
		elif command_name == "scenario-create":
			steps.append("Edit the generated scenario to add specific test cases")
			steps.append("Run scenario-evaluate to test the new scenario")
		elif command_name == "scenario-evaluate":
			steps.append("Run scenario-coverage to check test coverage")
		elif command_name == "scenario-coverage":
			steps.append("Use scenario-suggest to find gaps")
			steps.append("Create scenarios for untested commands")
	else:
		error = result.get("error")
		if error:
			suggestion = error.get("suggestion") if isinstance(error, dict) else getattr(error, "suggestion", None)
			if suggestion:
				steps.append(suggestion)

			code = error.get("code") if isinstance(error, dict) else getattr(error, "code", "")
			if code == "PARSE_ERROR":
				steps.append("Check scenario YAML syntax")
			elif code == "FILE_NOT_FOUND":
				steps.append("Verify the scenario file path exists")
			elif code == "VALIDATION_ERROR":
				steps.append("Review the scenario schema requirements")

	return steps


def _calculate_interpretation_confidence(result: dict[str, Any]) -> float:
	"""Calculate confidence in how accurately we can interpret the result."""
	confidence = 0.9

	if not result.get("success"):
		error = result.get("error")
		code = None
		if isinstance(error, dict):
			code = error.get("code")
		elif error:
			code = getattr(error, "code", None)
		if not code:
			confidence -= 0.2

	if result.get("reasoning"):
		confidence += 0.05

	sources = result.get("sources")
	if sources and len(sources) > 0:
		confidence += 0.05

	return max(0.0, min(1.0, confidence))
